pkg/component/autoscaling/gardenercustommetrics/component.go

```diff
@@ -0,0 +1,215 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package gardenercustommetrics implements the gardener-custom-metrics seed component (aka GCMx).
+// For details, see the GardenerCustomMetrics type.
+package gardenercustommetrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/component/autoscaling/gardenercustommetrics/kubeobjects"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	secretsutils "github.com/gardener/gardener/pkg/utils/secrets"
+	secretsmanager "github.com/gardener/gardener/pkg/utils/secrets/manager"
+)
+
+// ComponentName is the component name.
+const ComponentName = componentBaseName
+
+// GardenerCustomMetrics manages an instance of the gardener-custom-metrics component (aka GCMx). The component is
+// deployed on a seed, scrapes the metrics from all shoot kube-apiserver pods, and provides custom metrics by
+// registering as APIService at the custom metrics extension point of the seed kube-apiserver.
+// For information about individual fields, see the NewGardenerCustomMetrics function.
+type GardenerCustomMetrics struct {
+	namespaceName      string
+	containerImageName string
+	kubernetesVersion  *semver.Version
+
+	seedClient     client.Client
+	secretsManager secretsmanager.Interface
+}
+
+// NewGardenerCustomMetrics creates a new GardenerCustomMetrics instance tied to a specific server connection.
+//
+// namespace is where the server-side artefacts (e.g. pods) will be deployed (usually the 'garden' namespace).
+// containerImageName points to the binary for the gardener-custom-metrics pods. The exact version to be used, is
+// determined by contextual configuration, e.g. image vector overrides.
+// If enabled is true, this instance strives to bring the component to an installed, working state. If enabled is
+// false, this instance strives to uninstall the component.
+// seedClient represents the connection to the seed API server.
+// secretsManager is used to interact with secrets on the seed.
+func NewGardenerCustomMetrics(
+	namespace string,
+	containerImageName string,
+	kubernetesVersion *semver.Version,
+	seedClient client.Client,
+	secretsManager secretsmanager.Interface) *GardenerCustomMetrics {
```

Comment:
It's a definite improvement. I'll apply to similar cases, if I spot any in the code.

I didn't know that's the format we use in case of multiline function signature. My standard approach is to leave an empty line at the start of the body (example below), but that's rejected by G/G's lint configuration. Admittedly, with tab size = 4, G/G's way may actually be neater.
```
func New(
	namespace string,
	values Values,
	seedClient client.Client,
	secretsManager secretsmanager.Interface) component.DeployWaiter {

	return &gardenerCustomMetrics{
		namespaceName:  namespace,
		values:         values,
		client:         seedClient,
		secretsManager: secretsManager,
	}
}
```